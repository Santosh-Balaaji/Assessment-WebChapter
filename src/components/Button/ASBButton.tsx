import React from 'react';
import PropTypes from 'prop-types';
import Button from '@material-ui/core/Button';
import classNames from 'classnames';
import { IASBButtonProps } from '../../utils/interfaces';

const ASBButton: React.FC<IASBButtonProps> = ({
  children,
  color,
  className,
  minWidth,
  size,
  variant,
  onClick,
}): JSX.Element => {
  const colorSelection = ['success', 'tertiary', 'warning', 'info'];
  let customStyles = {};
  if (minWidth) {
    customStyles = { minWidth };
  }

  if (color && colorSelection.includes(color)) {
    return (
      <Button data-testid="StyledButton" style={customStyles} onClick={onClick}>
        {children}
      </Button>
    );
  }
  return (
    <Button
      data-testid="DefaultButton"
      color={color}
      size={size}
      variant={variant}
      className={classNames(className)}
      onClick={onClick}
    >
      {children}
    </Button>
  );
};

export default ASBButton;

ASBButton.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.node,
    PropTypes.func,
    PropTypes.string,
  ]).isRequired,
  color: PropTypes.oneOf(['default', 'inherit', 'primary', 'secondary']),
  size: PropTypes.oneOf(['small', 'medium', 'large']),
  variant: PropTypes.oneOf(['text', 'outlined', 'contained']),
  className: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.func,
    PropTypes.element,
  ]),
  minWidth: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  onClick: PropTypes.func.isRequired,
};

ASBButton.defaultProps = {
  color: 'primary',
  className: '',
  minWidth: undefined,
  size: 'medium',
  variant: 'outlined',
};
