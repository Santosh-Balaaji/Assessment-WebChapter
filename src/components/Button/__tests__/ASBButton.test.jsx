import React from 'react';
import { shallow } from 'enzyme';
import ASBButton from '../ASBButton';

describe('ASB Button Component', () => {
  let wrapper, wrapper1;
  const childrenName = 'Submit';
  const mockOnClick = jest.fn();
  const stubProps = {
    children: childrenName,
    color: 'primary',
    minWidth: 250,
    size: 'medium',
    variant: 'outlined',
    onClick: mockOnClick,
  };

  const mockProps = {
    children: childrenName,
    color: 'success',
    minWidth: 250,
    size: 'medium',
    variant: 'outlined',
    onClick: mockOnClick,
  };

  beforeAll(() => {
    wrapper = shallow(<ASBButton {...stubProps} />);
    wrapper1 = shallow(<ASBButton {...mockProps} />);
  });

  it('should render default component', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapper1).toMatchSnapshot();
  });

  it('should call handleClick function on click of Button', () => {
    wrapper
      .find('WithStyles(ForwardRef(Button))[data-testid="DefaultButton"]')
      .simulate('click');
    expect(mockOnClick).toHaveBeenCalled();
  });

  it('should set the name passed as children', () => {
    expect(
      wrapper
        .find('WithStyles(ForwardRef(Button))[data-testid="DefaultButton"]')
        .text(),
    ).toEqual(childrenName);
  });

  it('should display styled button when color is present in colorList', () => {
    expect(
      wrapper1
        .find('WithStyles(ForwardRef(Button))[data-testid="StyledButton"]')
        .text(),
    ).toEqual(childrenName);
  });
});
