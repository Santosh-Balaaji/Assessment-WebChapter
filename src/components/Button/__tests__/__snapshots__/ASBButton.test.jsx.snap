// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASB Button Component should render default component 1`] = `
<WithStyles(ForwardRef(Button))
  className=""
  color="primary"
  data-testid="DefaultButton"
  onClick={[MockFunction]}
  size="medium"
  variant="outlined"
>
  Submit
</WithStyles(ForwardRef(Button))>
`;

exports[`ASB Button Component should render default component 2`] = `
<WithStyles(ForwardRef(Button))
  data-testid="StyledButton"
  onClick={[MockFunction]}
  style={
    Object {
      "minWidth": 250,
    }
  }
>
  Submit
</WithStyles(ForwardRef(Button))>
`;
